### PR TITLE
test: clickhouse integration tests

### LINKED ISSUE #2813: ClickHouse integration test
I tried that with #2804 and #2811 and failed.
Perhaps refactoring to separate `import_csv` etc from `DbConnection` first as in #2804, then we can add ClickHouse as in #2811.

### PR DESCRIPTION
**Pull request by eitsupi**
Close #2813, based on #2814

Same as #2811

### CODE REVIEW

**Comment by eitsupi** [comment]
Oh, I see. The order of the Clickhouse results seems to be random.
https://stackoverflow.com/questions/54786494/clickhouse-query-row-order-behaviour

I have observed that the order of the results changes each time I run the following query.

```sql
WITH table_0 AS (
  SELECT
    13 AS x_int,
    13.0 AS x_float,
    5 AS k_int,
    5.0 AS k_float
  UNION
  ALL
  SELECT
    -13 AS x_int,
    -13.0 AS x_float,
    5 AS k_int,
    5.0 AS k_float
  UNION
  ALL
  SELECT
    13 AS x_int,
    13.0 AS x_float,
    -5 AS k_int,
    -5.0 AS k_float
  UNION
  ALL
  SELECT
    -13 AS x_int,
    -13.0 AS x_float,
    -5 AS k_int,
    -5.0 AS k_float
)
SELECT
  (x_int / k_int)
FROM
  table_0
```

@max-sixty @aljazerzen Do you think we can add a `sort` at the end of every PRQL query so that even systems like ClickHouse match the results?

File: prql-compiler/tests/integration/docker-compose.yml [hunk 1]
```diff
@@ -46,3 +46,18 @@ services:
...
+      # 9004 is MySQL emulation port
+      # https://clickhouse.com/docs/en/guides/sre/network-ports
+      - "9004:9004"
+    environment:
+      CLICKHOUSE_DB: dummy
+      # Skip `chown` to user_files_path
+      # https://github.com/ClickHouse/ClickHouse/blob/01c7d2fe719f9b9ed59fce58d5e9dec44167e42f/docker/server/entrypoint.sh#L7-L9
+      CLICKHOUSE_DO_NOT_CHOWN: "1"
+    volumes:
+      # ClickHouse can load csv only from user_files_path (default `/var/lib/clickhouse/user_files/`
```
Comment (eitsupi) [inline prql-compiler/tests/integration/docker-compose.yml]:
```suggestion
      # ClickHouse can load csv only from user_files_path (default `/var/lib/clickhouse/user_files/`)
```

**Comment by max-sixty** [comment]
> @max-sixty @aljazerzen Do you think we can add a `sort` at the end of every PRQL query so that even systems like ClickHouse match the results?

Yes I think that's OK — it doesn't limit any SQL functionality. Feel free to merge this with most queries excluded and then follow-up with changed queries unexcluded.

I also think we could have the Unsupported dialects look for e.g. `# run_clickhouse`, rather than the absence of e.g. `# skip_clickhouse`, if that lets us add queries without worrying about the tail of dialects. Supported dialects would still require explicit skipping.

(Though maybe that's harder than just adding `# skip_clickhouse` to the others, feel free to do whatever is easier)

File: prql-compiler/src/sql/dialect.rs [hunk 2]
```diff
@@ -81,11 +81,12 @@ impl Dialect {
             | Dialect::SQLite
             | Dialect::Postgres
             | Dialect::MySql
-            | Dialect::MsSql => SupportLevel::Supported,
+            | Dialect::MsSql
+            | Dialect::ClickHouse => SupportLevel::Supported,
```
Comment (max-sixty) [inline prql-compiler/src/sql/dialect.rs:85]: Depending on how many exclusions we need, we could move this down a rung. (Probably we should do that for MsSql too)

**Review by max-sixty — APPROVED** [review]
Great!

**Comment by eitsupi** [comment]
Note: ClickHouse doesn't have `lag`/`lead` functions.
https://clickhouse.com/docs/en/sql-reference/window-functions

**Comment by eitsupi** [comment]
The problem that we are currently experiencing is that the results of the following query.

https://github.com/PRQL/prql/blob/f35c7e5901552fdb9e617f6f61f712cf3fa588c0/prql-compiler/tests/integration/queries/switch.prql#L1-L8

This is not a connection problem, as the following SQL via MySQL CLI reproduces it.

```sql
WITH table_0 AS (
  SELECT
    CASE
      WHEN composer IS NOT NULL THEN composer
      WHEN genre_id < 17 THEN 'no composer'
      ELSE 'unknown composer'
    END AS display,
    milliseconds
  FROM
    tracks
  ORDER BY
    milliseconds
  LIMIT
    10
)
SELECT
  display
FROM
  table_0
ORDER BY
  milliseconds
```

```
+---------------+
| display       |
+---------------+
| Samuel Rosa   |
|               |
|               |
|               |
| L. Muggerud   |
|               |
| L. Muggerud   |
|               |
| Gilberto Gil  |
| Chico Science |
+---------------+
```

**Comment by eitsupi** [comment]
I expected it to be a case when issue, but the following example works.......

```elm
from [
    {x = '3'},
    {x = '1'},
]
select x = case {
    x == '3' => 'fiz',
    true => x,
}
```

```
MySQL [dummy]> WITH table_0 AS (
    ->   SELECT
    ->     '3' AS x
    ->   UNION
    ->   ALL
    ->   SELECT
    ->     '1' AS x
    -> )
    -> SELECT
    ->   CASE
    ->     WHEN x = '3' THEN 'fiz'
    ->     ELSE x
    ->   END AS x
    -> FROM
    ->   table_0;
+------+
| x    |
+------+
| fiz  |
| 1    |
+------+
2 rows in set (0.025 sec)
```

**Comment by eitsupi** [comment]
Oh, I understand. Perhaps this is being read as an empty character rather than `NULL`...

```sql
WITH table_0 AS (
  SELECT
    CASE
      WHEN composer != '' THEN composer
      WHEN genre_id < 17 THEN 'no composer'
      ELSE 'unknown composer'
    END AS display,
    milliseconds
  FROM
    tracks
  ORDER BY
    milliseconds
  LIMIT
    10
)
SELECT
  display
FROM
  table_0
ORDER BY
  milliseconds
```

```
+------------------+
| display          |
+------------------+
| Samuel Rosa      |
| no composer      |
| no composer      |
| no composer      |
| L. Muggerud      |
| no composer      |
| L. Muggerud      |
| unknown composer |
| Gilberto Gil     |
| Chico Science    |
+------------------+
```

**Comment by eitsupi** [comment]
~~https://clickhouse.com/docs/en/operations/settings/formats#format_csv_null_representation~~
-> Seems not related to...

It seems to be correctly recognized as NULL when reading csv. Perhaps a problem with the table type definition?
https://clickhouse.com/docs/en/sql-reference/data-types/nullable

```sql
> SELECT composer FROM file('/var/lib/clickhouse/user_files/chinook/tracks.csv') limit 5;
+------------------------------------------------------------------------+
| composer                                                               |
+------------------------------------------------------------------------+
| Angus Young, Malcolm Young, Brian Johnson                              |
| NULL                                                                   |
| F. Baltes, S. Kaufman, U. Dirkscneider & W. Hoffman                    |
| F. Baltes, R.A. Smith-Diesel, S. Kaufman, U. Dirkscneider & W. Hoffman |
| Deaffy & R.A. Smith-Diesel                                             |
+------------------------------------------------------------------------+
5 rows in set (0.010 sec)
```

**Comment by eitsupi** [comment]
Ok, `Nullable` works...

```sql
> WITH table_0 AS (
    ->   SELECT
    ->     CASE
    ->       WHEN composer IS NOT NULL THEN composer
    ->       WHEN genre_id < 17 THEN 'no composer'
    ->       ELSE 'unknown composer'
    ->     END AS display,
    ->     milliseconds
    ->   FROM
    ->     tracks
    ->   ORDER BY
    ->     milliseconds
    ->   LIMIT
    ->     10
    -> )
    -> SELECT
    ->   display
    -> FROM
    ->   table_0
    -> ORDER BY
    ->   milliseconds;
+------------------+
| display          |
+------------------+
| Samuel Rosa      |
| no composer      |
| no composer      |
| no composer      |
| L. Muggerud      |
| no composer      |
| L. Muggerud      |
| unknown composer |
| Gilberto Gil     |
| Chico Science    |
+------------------+
10 rows in set (0.011 sec)
```

**Comment by eitsupi** [comment]
It seems working now 🎉

**Comment by max-sixty** [comment]
Great work!!

**Pull request merged**